builder/tests/test_actions.py

```diff
@@ -90,60 +90,175 @@ def test_duplicate_search(version_from_scratch):
+        |                               |
+    - 439656005                     + 73583000
+    Arthritis of elbow            Epicondylitis
+        |                               |
+        -----------202855006-------------
+              - Lateral epicondylitis
+    """
+    # Epicondylitis codes are already included on the codelist; one is included, one is not
+    epicondylitis_included_code = "73583000"
+    epicondylitis_excluded_code = "202855006"
+    epicondylitis_codes = {epicondylitis_included_code, epicondylitis_excluded_code}
+    # Tennis toe code is not on the codelist
+    tennis_toe_code = "238484001"
+
+    # Arthritis of elbow is on the codelist, but not included
+    arthritis_of_elbow_code = "439656005"
+
+    assert epicondylitis_included_code in included_codes
+    assert epicondylitis_excluded_code not in included_codes
+    for code in epicondylitis_codes:
+        assert code in codelist_codes
+    assert tennis_toe_code not in codelist_codes
+    assert tennis_toe_code not in included_codes
+    assert arthritis_of_elbow_code in codelist_codes
+    assert arthritis_of_elbow_code not in included_codes
+
+    # Act: create a term search
+    s = actions.create_search(
+        draft=draft,
+        term="Tennis toe",
+        codes={tennis_toe_code},
+    )
+
+    # Act: delete the search
+    actions.delete_search(search=s)
+
+    # Assert...
+    # that the codelist has 0 searches
+    assert draft.searches.count() == 0
+    # that the codelist still has the codes that were not associated with the search
+    assert draft.code_objs.count() == len(codelist_codes)
+
+    # Arrange: create new searches
+    s1 = actions.create_search(
+        draft=draft,
+        term="epicondylitis",
+        codes=epicondylitis_codes,
+    )
+    s2 = actions.create_search(draft=draft, term="Tennis toe", codes={tennis_toe_code})
+
+    # Assert...
+    # that the codelist now has 2 searches
+    assert draft.searches.count() == 2
+    # and one extra code obj
+    assert draft.code_objs.count() == len(codelist_codes) + 1
+
+    # Act: delete the search for "Tennis toe"
+    actions.delete_search(search=s2)
+
+    # Assert...
+    # that the codelist has only 1 search
+    assert draft.searches.count() == 1
+    # that the codelist still has the starting codes that were not associated with the search
+    assert draft.code_objs.count() == len(codelist_codes)
+
+    # Act: delete the search for "epicondylitis"
+    actions.delete_search(search=s1)
+
+    # Assert...
+    # that the codelist has no searches
+    assert draft.searches.count() == 0
+    # This search matched 2 codes, one included, one not included, but a descendant of an included
+    # code, they are not deleted
+    assert draft.code_objs.count() == len(codelist_codes)
+
+    # Arrange: create new search for not-included code
+    s3 = actions.create_search(
+        draft=draft,
+        term="arthritis of elbow",
+        codes={arthritis_of_elbow_code},
```

Comment:
This isn't quite right, although it's almost certainly fine.

The view code that creates a search passes all the codes that are found by [`codelists.search.do_search`](https://github.com/opensafely-core/opencodelists/blob/main/codelists/search.py) -- so not just the codes that match, but their descendants too.

https://github.com/opensafely-core/opencodelists/blob/5e46909d7ae2b8e2a112435029852978200281d4/builder/views.py#L212-L214

So in this case, the `codes` argument should also include the "lateral epicondylitis" code, since that's a descendant of "arthritis of elbow".

Reply:
updated